framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+     * format, all with appropriate annotations.
+     *
+     * @param scene the scene to write
+     * @param printWriter where to write the scene to
+     */
+    private void writeImpl(AScene scene, PrintWriter printWriter) {
+        // Write out all imports
+        ImportDefCollector importDefCollector;
+        try {
+            importDefCollector = new ImportDefCollector(scene, printWriter);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+        importDefCollector.visit();
+        printWriter.println();
+
+        // For each class
+        for (Map.Entry<String, AClass> classEntry : scene.classes.entrySet()) {
+            @SuppressWarnings("signature") // TODO: annotate AScene library
+            @BinaryName String classname = classEntry.getKey();
+            AClass aClass = classEntry.getValue();
+            String pkg = packagePart(classname);
+            String basename = basenamePart(classname);
+
+            // Do not attempt to print stubs for anonymous inner classes, because the stub parser
+            // cannot read them. (An anonymous inner class has a basename like Outer.1, so this
+            // check ensures that the binary name's final segment after its last . is not only
+            // composed of digits.)
+            if (basename.contains(".")) {
+                String innermostClassname = basename.substring(basename.lastIndexOf('.') + 1);
+                if (innermostClassname.matches("\\d+")) {
+                    continue;
+                }
+            }
+
+            if (!"".equals(pkg)) {
+                printWriter.println("package " + pkg + ";");
+            }
+            if ("package-info".equals(basename) || "module-info".equals(basename)) {
+                continue;
+            }
+
+            // At this point, we no longer care about the distinction between packages
+            // and inner classes, so we should replace the $ in the definition of any
+            // inner classes with a ., so that they are printed correctly in stub files.
+
+            String fullyQualifiedClassname = Signatures.binaryNameToFullyQualified(classname);
+
+            int curlyCount =
+                    1
+                            + printClassDefinitions(
+                                    basename, fullyQualifiedClassname, aClass, printWriter);
+
+            // print fields or enum constants
+            if (!enumConstants.containsKey(fullyQualifiedClassname)) {
+                printFields(aClass, fullyQualifiedClassname, printWriter);
+            } else {
+                // for enums, instead of printing fields print the enum constants
+                List<VariableElement> enumConstants =
+                        this.enumConstants.get(fullyQualifiedClassname);
+
+                StringJoiner sj = new StringJoiner(", ");
+                for (VariableElement enumConstant : enumConstants) {
+                    sj.add(enumConstant.getSimpleName());
```

Comment:
Enum constants are allowed to be annotated.  Could you add a test for that?

Reply:
WPI doesn't currently have any way to annotate enum constants, because they can't be assigned to. I've added a (skipped) test in `EnumConstants.java`, but I'm not sure how useful it is. Both stub and jaif WPI fail it, right now.